components/cli/pkg/registry/credentials/credentials.go

```diff
@@ -0,0 +1,187 @@
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package credentials
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"golang.org/x/crypto/ssh/terminal"
+
+	"github.com/cellery-io/sdk/components/cli/pkg/config"
+	"github.com/cellery-io/sdk/components/cli/pkg/util"
+)
+
+// FromBrowser requests the credentials from the user
+func FromBrowser(username string) (string, string, error) {
+	auth := config.LoadConfig().AuthConf
+	timeout := make(chan bool)
+	ch := make(chan string)
+	var code string
+	httpPortString := ":" + strconv.Itoa(auth.CallBackDefaultPort)
+	var codeReceiverPort = auth.CallBackDefaultPort
+	// This is to start the CLI auth in a different port is the default port is already occupied
+	for {
+		_, err := net.Dial("tcp", httpPortString)
+		if err != nil {
+			break
+		}
+		codeReceiverPort++
+		httpPortString = ":" + strconv.Itoa(codeReceiverPort)
+	}
+	redirectUrl := url.QueryEscape("http://" + auth.CallBackHost + ":" + strconv.Itoa(codeReceiverPort) + "/auth")
+	var gitHubAuthUrl = "https://" + auth.IsHost + ":" + strconv.Itoa(auth.IsPort) +
+		"/oauth2/authorize?scope=openid&response_type=" + "code&redirect_uri=" + redirectUrl + "&client_id=" +
+		auth.SpClientId + "&fidp=google"
+
+	fmt.Println("\n", gitHubAuthUrl)
+	go func() {
+		mux := http.NewServeMux()
+		server := http.Server{Addr: httpPortString, Handler: mux}
+		//var timer *time.Timer
+		mux.HandleFunc(auth.CallBackContextPath, func(w http.ResponseWriter, r *http.Request) {
+			err := r.ParseForm()
+			if err != nil {
+				util.ExitWithErrorMessage("Error parsing the code", err)
+			}
+			code = r.Form.Get(auth.CallBackParameter)
+			ch <- code
+			if len(code) != 0 {
+				http.Redirect(w, r, "https://"+auth.IsHost+":"+strconv.Itoa(auth.IsPort)+
+					"/authenticationendpoint/auth_success.html", http.StatusSeeOther)
+			} else {
+				util.ExitWithErrorMessage("Did not receive any code", err)
+			}
+			flusher, ok := w.(http.Flusher)
+			if !ok {
+				util.ExitWithErrorMessage("Error in casting the flusher", err)
+			}
+			flusher.Flush()
+			err = server.Shutdown(context.Background())
+			if err != nil {
+				util.ExitWithErrorMessage("Error while shutting down the server\n", err)
+			}
+		})
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			util.ExitWithErrorMessage("Error while establishing the service", err)
+		}
+	}()
+
+	err := util.OpenBrowser(gitHubAuthUrl)
```

Comment:
Shouldn't this be done before starting the server ? Because if this is run in a headless server, the HTTP server is not required.

Reply:
According to the requirement we need to start a server before starting the browser.